devices/MulticastDNS/Package/PacketParser.cs

```diff
@@ -0,0 +1,156 @@
+// The .NET Foundation licenses this file to you under the MIT license.
+
+using System;
+using System.Text;
+
+namespace Iot.Device.MulticastDNS.Package
+{
+    internal class PacketParser
+    {
+        private readonly IBitConverter _converter = EndianBitConverter.Big;
+        private readonly Encoding _encoding = Encoding.UTF8;
+        private readonly byte[] _data;
+        private int _position = 0;
+
+        public PacketParser(byte[] bytes)
+        {
+            _data = bytes;
+        }
+
+        private void MovePosition(int n)
+        {
+            _position += n;
+
+            if (_position > _data.Length)
+            {
+                throw new IndexOutOfRangeException("No more data in packet");
+            }
+        }
+
+        public byte ReadByte() => _data[_position++];
+
+        public ushort ReadUShort()
+        {
+            ushort value = _converter.ToUInt16(_data, _position);
+            MovePosition(2);
+            return value;
+        }
+
+        public int ReadInt()
+        {
+            int value = _converter.ToInt32(_data, _position);
+            MovePosition(4);
+            return value;
+        }
+
+        public byte[] ReadBytes(int count)
+        {
+            byte[] value = new byte[count];
+            System.Array.Copy(_data, _position, value, 0, count);
+            MovePosition(count);
+            return value;
+        }
+
+        public string ReadString()
+        {
+            int length = ReadByte();
+            int remaining = _data.Length - _position;
+
+            if (length > remaining)
+            {
+                length = remaining;
+            }
+
+            string value = _encoding.GetString(_data, _position, length);
+            MovePosition(length);
+            return value;
+        }
+
+        public string ReadDomain()
+        {
+            int dot = 0;
+            string domain = string.Empty;
+
+            while (true)
+            {
+                string label = PopLabel(ref dot);
+
+                if (label == null)
+                {
+                    break;
+                }
+
+                domain += label + ".";
+            }
+
+            return domain.Trim('.');
+        }
+
+        private string PopLabel(ref int dot)
+        {
+            int length = PopLabelLength(ref dot);
+            if (length == 0)
+            {
+                return null;
+            }
+
+            if (dot != 0)
+            {
+                return GetPointer(length, ref dot);
+            }
+
+            string label = _encoding.GetString(_data, _position, length);
+            MovePosition(length);
+
+            return label;
+        }
+
+        private int PopLabelLength(ref int dot)
+        {
+            if (dot != 0)
+            {
+                return GetPointerLength(ref dot);
+            }
+
+            int length = _data[_position++];
+            if ((length & 0xc0) != 0xc0)
```

Comment:
a little comment on those magic numbers would be perfect.

Reply:
Done